### PR TITLE
EVEREST-1054 | Fix PG restores for S3 compatible storage locations 

### PR DESCRIPTION
**Pull request by mayankshah1607**
**CHANGE DESCRIPTION**
---
**Problem:**
EVEREST-1054

PG restores do not work when using a backup storage location such as Minio.

**Cause:**
The restore issues are only caused when setting `forcePathStyle` and/or `verifyTLS=false`

#### 1. When restoring to same cluster:
The backup controller lists objects in the S3 storage to build the backup destination path. This S3 API call was failing as the client was not configured for skipping TLS checks or using path-style URIs.

#### 2. When restoring to new cluster:
When performing restores to a new cluster, we did not configure pgBackRest to skip TLS checks or force path-style URIs, depending on the backup storage.

**Solution:**
- Enforce TLS settings and URI style in S3 client while listing S3 objects in the backup controller
- Similarly, configure pgBackRest settings correctly when restoring to a new cluster.

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?

**Tests**
- [x] Is an Integration test/test case added for the new feature/change?
- [x] Are unit tests added where appropriate?

### CODE REVIEW

File: controllers/providers/pg/applier.go [hunk 1]
```diff
@@ -588,11 +588,21 @@ func (p *applier) genPGDataSourceSpec() (*crunchyv1beta1.DataSource, error) {
...
+	backupBaseName, repoName string) ([]string, error) {
 	options := []string{
 		"--set=" + backupBaseName,
 	}
 
+	if pointer.Get(backupStorage.Spec.ForcePathStyle) {
+		options = append(options, "--"+fmt.Sprintf(pgBackRestStorageForcePathTmpl, repoName)+"="+pgBackRestStoragePathStyle)
+	}
+	if !pointer.Get(backupStorage.Spec.VerifyTLS) {
+		options = append(options, "--no-"+fmt.Sprintf(pgBackRestStorageVerifyTmpl, repoName))
```
Comment (recharte) [inline controllers/providers/pg/applier.go:601]: Are you sure this works? I couldn't find this in the PGBackrest docs...
I could only find `repo1-storage-verify-tls=n`
https://pgbackrest.org/configuration.html#section-repository/option-repo-storage-verify-tls

File: controllers/providers/pg/applier.go [hunk 2]
```diff
@@ -586,11 +586,21 @@ func (p *applier) genPGDataSourceSpec() (*crunchyv1beta1.DataSource, error) {
...
+	}
```
Comment (recharte) [inline controllers/providers/pg/applier.go:602]: I'm a bit confused why we need this at all. Should [this](https://github.com/percona/everest-operator/blob/939c65dc1771cef0b9ef2a645f173fa15f219809/controllers/providers/pg/applier.go#L1116-L1123) already take care of it? :thinking:

File: controllers/providers/pg/applier.go (hunk 1 above)
Comment (mayankshah1607) [reply controllers/providers/pg/applier.go:601]: I wasn't able to get the anchor tag for this, but you'll find it under this section: https://pgbackrest.org/command.html#command-restore

Screenshot 2024-05-08 at 6 36 33 PM

File: controllers/providers/pg/applier.go (hunk 2 above)
Comment (mayankshah1607) [reply controllers/providers/pg/applier.go:602]: That's what I thought at first, but it doesn't. The snippet you provided only takes care of the backups.
When you restore to a new cluster, those settings are lost. So you need to set them again

Comment (mayankshah1607) [reply controllers/providers/pg/applier.go:603]: This `getPGRestoreOptions` function is called only when we specify the `DataSource` field. Whereas in the snipped provided above, those are used only for backups and restoring to the same cluster. So I don't think there should be any conflicts/issues..

Comment (mayankshah1607) [reply controllers/providers/pg/applier.go:603]: BTW, I did a run through of the backup/restore scenario with these changes, and it worked as expected

Comment (recharte) [reply controllers/providers/pg/applier.go:603]: The thing that's concerning me is the hardcoded `repo1` (I know it was already like this).

Comment (recharte) [reply controllers/providers/pg/applier.go:603]: Okay so I just dug a bit deeper on this and I think everything is fine. Like you said this is completely isolated from the actual DB. The repos that we define here with the `handlePGDataSourceS3` have absolutely nothing to do with the ones defined outside the `DataSource` field.

**Review by recharte — APPROVED** [review]

**Review by oksana-grishchenko — APPROVED** [review]

**Pull request merged**